src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5957,13 +5957,14 @@ private ScanTypeFlags ScanNonArrayType(ParseTypeMode mode, out SyntaxToken lastT
                 lastTokenOfType = this.EatToken();
                 result = ScanTypeFlags.MustBeType;
             }
-            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken && mode != ParseTypeMode.DefinitePattern)
+            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
             {
                 lastTokenOfType = this.EatToken();
 
                 result = this.ScanTupleType(out lastTokenOfType);
-                if (result == ScanTypeFlags.NotType)
+                if (result == ScanTypeFlags.NotType || mode == ParseTypeMode.DefinitePattern && this.CurrentToken.Kind != SyntaxKind.OpenBracketToken)
```

Comment:
The name `ScanNonArrayType` makes it sound like the precondition for calling this method is that the type is not an array. Is that not true?

Reply:
It is, but only trivially so. Every array type starts with a non-array type. This code scans that non-array type. It refuses to scan the array brackets and just leaves them as the next tokens on the input.

---
In reply to: [181215276](https://github.com/dotnet/roslyn/pull/26088#discussion_r181215276) [](ancestors = 181215276)